compiler/src/iree/compiler/GlobalOptimization/PropagateLinalgTranspose.cpp

```diff
@@ -0,0 +1,602 @@
+    // Becomes
+    //
+    // reassociation_map = [[0, 1, 2], [3, 4], [5]]
+    // permutation = [0, 1, 2, 4, 5, 3]
+    applyPermutationToVector(reassociations, invPerm);
+
+    SmallVector<int64_t> newInvPerm;
+    SmallVector<ReassociationIndices> newReassociations;
+    int64_t expandedDim = 0;
+    for (auto reassoc : reassociations) {
+      ReassociationIndices newReassoc;
+      for (auto dim : reassoc) {
+        newInvPerm.push_back(dim);
+        newReassoc.push_back(expandedDim++);
+      }
+      newReassociations.push_back(newReassoc);
+    }
+
+    auto newPerm = invertPermutationVector(newInvPerm);
+
+    RankedTensorType expandedType = getPermutedTensorType(
+        cast<RankedTensorType>(expandOp.getType()), newInvPerm);
+    Value transposedReshape = rewriter.create<tensor::ExpandShapeOp>(
+        expandOp.getLoc(), expandedType, transposeOp.getInput(),
+        newReassociations);
+    Value originalReshape =
+        createTranspose(rewriter, transposedReshape, newPerm);
+    rewriter.replaceOp(expandOp, originalReshape);
+    return success();
+  }
+};
+
+// Fuses a transpose with the input of a linalg.generic op. For example,
+//
+// %0 = linalg.transpose ... permutation = [0, 2, 1] :
+//                           tensor<2x5x7> -> tensor<2x7x5>
+// linalg.generic {
+//   indexing_maps = [affine_map<(d0, d1, d2) -> (d1, d0, d2)>,
+//                    affine_map<(d0, d1, d2) -> (d0, d1, d2)>]
+//   ins(%0 : tensor<2x7x5>)
+//
+// Becomes
+//
+// linalg.generic {
+//   indexing_maps = [affine_map<(d0, d1, d2) -> (d1, d2, d0)>,
+//                    affine_map<(d0, d1, d2) -> (d0, d1, d2)>]
+//   ins(%0 : tensor<2x5x7>)
+//
+// This is considered just one way to model transpose propagation to generics,
+// another option would be to interpret the transpose on the iterators of the
+// generic, thus producing a transpose on the output and any other inputs to
+// the generic. This has the potential introduce more transposes/data movement
+// and isn't the way this pass is modeled. Global data layout transformations
+// like that are better suited for pack/unpack propagation rooted on specific
+// operations.
+class FuseTransposeWithGenericConsumer
+    : public OpRewritePattern<linalg::GenericOp> {
+public:
+  using OpRewritePattern<linalg::GenericOp>::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(linalg::GenericOp genericOp,
+                                PatternRewriter &rewriter) const override {
+    if (!IREE::Flow::isNonNullAndOutsideDispatch(genericOp)) {
+      return failure();
+    }
+    OpOperand *transposeOperand = nullptr;
+    linalg::TransposeOp transposeOp;
+    for (int64_t i = 0, e = genericOp.getInputs().size(); i < e; ++i) {
+      OpOperand *inputOperand = &genericOp->getOpOperand(i);
+      Value input = inputOperand->get();
+      transposeOp = input.getDefiningOp<linalg::TransposeOp>();
+      if (transposeOp) {
+        transposeOperand = inputOperand;
+        break;
+      }
+    }
```

Comment:
I think we can iterate over `getDpsInputOperands`, and simplify the code. E.g.,

```suggestion
    for (OpOperand* input : linalgOp.getDpsInputOperands()) {
      if (input->get().getDefiningOp<linalg::TransposeOp>()) {
        transposeOp = input;
        break
      }
    }
```